Refactor bootstrap+sandbox into "task" framework (#19275)
storybookjs/storybook: docs/contribute/code.md

git clone https://github.com/your-username/storybook.git
```

## Run your first sandbox

Storybook development happens in a set of *sandboxes* which are templated Storybook environments corresponding to different user setups. Within each sandbox, we inject a set of generalized stories that allow us to test core features and addons in all such environments.

To run an sandbox locally, you can use the `start` command:

```shell
yarn start
```

That will install the required prerequisites, build the code, create and link an example for a Vite/React setup, and start the storybook server.

If all goes well you should see the sandbox running.

> ![Storybook Sandbox Running](./storybook-cra-sandbox.png)

Reviewer: maybe rename the file to storybook-sandbox?

## Running a different sandbox template

The `start` command runs a Vite/React template, but there are many others you can use if you want to work on a different renderer or framework.

To get started, run the `yarn task` command; it will prompt you with a series of questions to figure out what you are trying to do. Once you've made a selection it will provide a set of options that you can use to run the same command again.

```shell
yarn task
```

<div class="aside">

> 💡 The <code>yarn task</code> command takes a couple of shortcuts for development that could catch you out if you change branches: you may need to rerun the <code>install</code> and <code>compile</code> tasks. You can do that with <code>--start-from=install</code></li>.

Reviewer: ```suggestion
💡 The <code>yarn task</code> command takes a couple of shortcuts for development that could catch you out if you change branches: you may need to rerun the <code>install</code> and <code>compile</code> tasks. You can do that by running the command with the <code>--start-from=install</code></li> flag.
```

</div>

## Running tests

Once you've run your [first sandbox](#run-your-first-sandbox), you should have a fully functional version of Storybook built on your local machine. Before making any code changes, it's helpful to verify that everything is working as it should. More specifically, the test suite.

Run the following command to execute the tests:

```shell
yarn test
```

## Start developing

> Now that you've [verified your setup](#sanity-check), it's time to jump into code. The simplest way to do this is to run one of the sandboxes in one terminal window and the interactive build process in a separate terminal.

Reviewer: ```suggestion
Now that you've [verified your setup](#running-tests), it's time to jump into code. The simplest way to do this is to run one of the sandboxes in one terminal window and the interactive build process in a separate terminal.
```

### Add stories

Adding a story or set of stories to our suite of generic stories helps you test your work.

> If you're modifying part of Storybook's core, or one of the essential addons, there's probably an existing set of stories in that addons `template/stories` folder that documents how the feature is supposed to work. Add your stories there.

Reviewer: ```suggestion
If you're modifying part of Storybook's core, or one of the essential addons, there's probably an existing set of stories in that addon's `template/stories` folder that documents how the feature is supposed to work. Add your stories there.
```

> If you're modifying something related to a specific renderer, the renderer will have its own template stories.

Reviewer: ```suggestion
If you're modifying something related to a specific renderer (e.g. React, Vue, etc.), the renderer will have its own template stories.
```

### End-to-end tests (e2e)

Storybook's monorepo is set up to rely on end-to-end testing with [Playwright](https://playwright.dev) during CI. To help with testing, we encourage running this test suite before submitting your contribution. 

To run a e2e test against a sandbox, you can use the `e2e-tests` task:

```shell
yarn task --task e2e-tests --template=<template> --start-from=auto

> ```

Reviewer: Might be possible that when people read this they might not know what `template` to fill in. Maybe specify `react-vite/default-ts` as an example, given it's the one coming from `yarn start`?

Reviewer: Feature request: It would be nice if we could run the commands without specifying a template and it would prompt. For instance, running `yarn task --task e2e-tests --start-from=auto` without `--template` results in the following error:
<img width="657" alt="image" src="https://user-images.githubusercontent.com/1671563/194706806-0da496d8-29f0-4df4-9555-e96ee60d3b38.png">

Author: Done!

- [Sync a fork](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/syncing-a-fork)
- [Merge an upstream repository into your fork](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/merging-an-upstream-repository-into-your-fork)

### Reproducing job failures

> If one of your CI jobs fails, it will print a message explaining how to reproduce the task locally. Typically that involves running the task against the right template:

Reviewer: ```suggestion
After creating your PR, if one of the CI jobs failed, when checking the logs of that job, you will see that it printed a message explaining how to reproduce the task locally. Typically that involves running the task against the right template:
```

```shell

> yarn task --task e2e-tests --template=cra-default-ts --start-from=install

Reviewer: Not sure if you forgot to change this to `react-vite/default-ts`, but if cra was intended, then it should have a slash

```suggestion
yarn task --task e2e-tests --template=cra/default-ts --start-from=install
```

Reviewer: Feature request: probably a good idea to also accept fully dashed template names, e.g. `react-vite-default-ts`, as it's how the folder names inside of `sandbox` are created, thus it's easier for contributors to understand how to find the template names.

This might not be 100% necessary though it'd be nice, however if users supply a wrong template they do get a nice feedback with the right options:
<img width="660" alt="image" src="https://user-images.githubusercontent.com/1671563/194707057-df82c9ca-9a1d-4ad9-9932-9e30b5aee89b.png">

Author: Yeah, there is an idea to allow either `--template=react-vite/default-js` OR `--sandbox=react-vite-default-js` (or both, in which case the `sandbox` step would create the project in the directory specified. I decided to leave that off this PR for now as it isn't really needed yet.

```

Typically it is a good idea to start from the `install` task to ensure your local code is completely up to date. If you reproduce the failure, you can try and make fixes, [compile them](#start-developing) with `build`, then rerun the task with `--start-from=auto`.
 
 <div class="aside">

> <p>💡 The default instructions run the code "linked" which means built changes to Storybook code will be reflected in the sandbox right away (the next time you run the task). However CI runs in "unlinked" mode, which in rare cases will behave differently.</p>

Reviewer: Could it be confusing for people to read "changes to Storybook code" ? Could they think like "does that mean changing the stories file of the example?" instead of "does that mean changing the source code of the Storybook libraries"

Author: ```suggestion
<p>💡 The default instructions run the code "linked" which means built changes to Storybook library code will be reflected in the sandbox right away (the next time you run the task). However CI runs in "unlinked" mode, which in rare cases will behave differently.</p>
```

> <p>If you are having trouble reproducing, try adding the <code>--no-link</code> flag. If you need to do that, you'll need to <code>--start-from=compile</code> after each code change.

Reviewer: ```suggestion
<p>If you are having trouble reproducing, try rerunning the command with the <code>--no-link</code> flag. If you need to do that, you'll need to run it with <code>--start-from=compile</code> after each code change.
```
